### ISSUE TITLE
Support for Anthropic Prompt Caching

### ISSUE THREAD

**Issue by danielflaherty** [Mirascope/mirascope#427]
### Description

Would be great if support for Anthropic's new prompt caching beta could be implemented:

https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching

Thanks!

**Comment by willbakst** [comment]
I definitely want to add support for this and will be looking into it.

I've added this to our v1.1 milestone!

**Comment by danielflaherty** [comment]
Great, thanks! Any rough idea of when 1.1 will be out? @willbakst

**Comment by willbakst** [comment]
Not certain yet (maybe a week or two?)

In the meantime, here is an example of using custom messages and headers to access the feature in v1:

```python
from mirascope.core import anthropic
import requests


@anthropic.call(
    "claude-3-5-sonnet-20240620",
    call_params={
        "max_tokens": 1024,
        "extra_headers": {"anthropic-beta": "prompt-caching-2024-07-31"},
    },
)
def analyze_book(book: str) -> anthropic.AnthropicDynamicConfig:
    return {
        "messages": [
            {
                "role": "system",
                "content": [
                    {
                        "type": "text",
                        "text": "You are an AI assistant tasked with analyzing literary works. Your goal is to provide insightful commentary on themes, characters, and writing style.\n",
                    },
                    {
                        "type": "text",
                        "text": f"{book}",
                        "cache_control": {"type": "ephemeral"},
                    },
                ],
            },
            {
                "role": "user",
                "content": "Analyze the major themes in Pride and Prejudice.",
            },
        ]
    }


response = requests.get("https://www.gutenberg.org/files/1342/1342-0.txt")
pride_and_prejudice = response.text

analysis = analyze_book(pride_and_prejudice)
print(analysis.content)
```

**Comment by willbakst** [comment]
Before working on an implementation, I'm trying to think through the design here. What do we think about using tags similar to `:image` for cached data? For example, maybe something like this?

```python
from mirascope.core import anthropic, prompt_template


@anthropi.call("claude-3-5-sonnet-20240620")
@prompt_template(
    """
    SYSTEM:
    You are an AI assistant tasked with analyzing literary works.
    Your goal is to provide insightful commentary on themes, characters, and writing style.
    
    {book:cache(type=ephemeral)}

    USER: Analyze the major themes in {title}
    """
)
def analyze_book(book: str, title: str): ...


book = requests.get("https://www.gutenberg.org/files/1342/1342-0.txt").text
analysis = analyze_book(book, "Pride and Prejudice")
print(analysis.content)
```

**Comment by danielflaherty** [comment]
Yeah that's the most natural way to express cache breakpoints.

One thing I wanted to mention is that I think it's also important to expose some of Anthropic's caching details. For example, breakpoints only work if the number of tokens preceding them exceeds some number (1024 in Sonnet 3.5's case), and you can only set a max of 4 breakpoints. It would be nice to raise a warning or something if the breakpoints did not hit. Or perhaps, could there be like an 'automatic' mode that sets the breakpoints for me, say after every 1024 tokens?

**Comment by willbakst** [comment]
Oh interesting, after some more digging I'm realizing we also need to consider how to cache tools.

I'm thinking something like this might be the best thing:

```python
from mirascope.core import BaseTool, ConfigDict, anthropic, prompt_template


class FormatAnalysis(BaseTool):
    analysis: str
    key_points: list[str]

    model_config = ConfigDict(cache_control={"type": "ephemeral"})

    def call(self) -> str:
        return f"Key Points:\n{self.key_points}\nFull Analysis:\n{self.analysis}"


@anthropic.call("claude-3-5-sonnet-20240620", tools=[FormatAnalysis])
@prompt_template(
    """
    SYSTEM:
    You are an AI assistant tasked with analyzing literary works.
    Your goal is to provide insightful commentary on themes, characters, and writing style.

    Here is the literary work, {title}: {book:cache_control(type=ephemeral)}

    USER: Analyze the major themes in {title}
    """
)
def analyze_book(book: str, title: str): ...


book = requests.get("https://www.gutenberg.org/files/1342/1342-0.txt").text
analysis = analyze_book(book, "Pride and Prejudice")
print(analysis.content)
```

The tools would then cache up to the breakpoint of the tool with the cache (like they have in their examples). This would also then enable caching to work for things like `response_model`.

For the breakpoint maximum, that's easily included (we just see how many breakpoints are included and raise a warning or error if it's more than the limit).

For the token limit stuff, I'm not sure if including a tokenizer and checking that is worthwhile compared to using their existing response booleans for whether or not a cache was created (i.e. there's no real difference in my book between counting and warning up-front vs. finding out after making the call that you were already prepared to make anyway). We can always show a warning based on this response boolean.

I will work to implement this soon hopefully, just need to finalize the design and implementation details :)

**Comment by willbakst** [comment]
This is released in `v1.1`!

docs: https://mirascope.io/docs/latest/learn/provider_specific_features/#prompt-caching
